### ISSUE TITLE
Configure non-nullable types as required

### ISSUE THREAD

**Issue by coder925** [domaindrivendev/Swashbuckle.AspNetCore#2036]
In my project we do not distinguish between `required` and `nullable: false`. If a type is non-nullable, it is also required. This is regardless if it is a value type or reference type.

It would be great if one could optionally configure Swashbuckle to set all non-nullable properties to required. E.g. like: `setupAction.MarkNonNullableTypesAsRequired();` This way, I could remove all my `[Required]` annotations on the non-nullable C# properties.

This is a subsequent request following  #1686. As _originally posted by @felschr in https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/1686#issuecomment-674074694_ , treating non-nullable as required is the default for [nullable contexts](https://docs.microsoft.com/en-us/aspnet/core/mvc/models/validation?view=aspnetcore-3.1#non-nullable-reference-types-and-required-attribute) in .NET validation.

**Comment by flibustier7seas** [comment]
Temprory solution:

```csharp
class RequiredNotNullableSchemaFilter : ISchemaFilter {
    public void Apply(OpenApiSchema schema, SchemaFilterContext context) {
        if (schema.Properties == null) {
            return;
        }

        var notNullableProperties = schema
            .Properties
            .Where(x => !x.Value.Nullable  && !schema.Required.Contains(x.Key))
            .ToList();

        foreach (var property in notNullableProperties)
        {
            schema.Required.Add(property.Key);
        }
    }
}
```

**Comment by JosNun** [comment]
@flibustier7seas you mention that your solution isn't completely correct, and I see you mention the phone field being marked as required, even though it's not. Do you know what in particular causes this case, or if there's any workarounds?

**Comment by flibustier7seas** [comment]
@JosNun I'm using the following workaround:

```csharp
class RequiredNotNullableSchemaFilter : ISchemaFilter
{
    public void Apply(OpenApiSchema schema, SchemaFilterContext context)
    {
        if (schema.Properties == null)
        {
            return;
        }

        FixNullableProperties(schema, context);

        var notNullableProperties = schema
            .Properties
            .Where(x => !x.Value.Nullable && !schema.Required.Contains(x.Key))
            .ToList();

        foreach (var property in notNullableProperties)
        {
            schema.Required.Add(property.Key);
        }
    }

    /// <summary>
    /// Option "SupportNonNullableReferenceTypes" not working with complex types ({ "type": "object" }), 
    /// so they always have "Nullable = false",
    /// see method "SchemaGenerator.GenerateSchemaForMember"
    /// </summary>
    private static void FixNullableProperties(OpenApiSchema schema, SchemaFilterContext context)
    {
        foreach (var property in schema.Properties)
        {
            if (property.Value.Reference != null)
            {
                var field = context.Type
                    .GetMembers(BindingFlags.Public | BindingFlags.Instance)
                    .FirstOrDefault(x =>
                        string.Equals(x.Name, property.Key, StringComparison.InvariantCultureIgnoreCase));

                if (field != null)
                {
                    var isReferenceType = field switch
                    {
                        FieldInfo fieldInfo => !fieldInfo.FieldType.IsValueType,
                        PropertyInfo propertyInfo => !propertyInfo.PropertyType.IsValueType,
                        _ => false,
                    };

                    if (isReferenceType)
                    {
                        property.Value.Nullable = !field.IsNonNullableReferenceType();
                    }
                }
            }
        }
    }
}
```

Result:

```json
{
  "User": {
    "required": ["email", "firstName", "id", "lastName"],
    "type": "object",
    "properties": {
      "id": {
        "type": "string",
        "format": "uuid"
      },
      "organizationId": {
        "type": "string",
        "format": "uuid",
        "nullable": true
      },
      "firstName": {
        "type": "string"
      },
      "middleName": {
        "type": "string",
        "nullable": true
      },
      "lastName": {
        "type": "string"
      },
      "email": {
        "$ref": "#/components/schemas/UserEmail"
      },
      "phone": {
        "$ref": "#/components/schemas/UserPhone"
      }
    },
    "additionalProperties": false
  }
}
```

**Comment by ahmednfwela** [comment]
small modification on @flibustier7seas 's answer
```cs
var notNullableProperties = schema
      .Properties
      .Where(x => !x.Value.Nullable && x.Value.Default == default && !schema.Required.Contains(x.Key))
      .ToList();
```

**Comment by brather1ng** [comment]
If you use the `SwaggerGenOptions.UseAllOfToExtendReferenceSchemas()` extension method, the workaround of @flibustier7seas is not necessary. With that, references are wrapped in an allOf and `SupportNonNullableReferenceTypes` also applies to them.

**Comment by flibustier7seas** [comment]
> If you use the `SwaggerGenOptions.UseAllOfToExtendReferenceSchemas()` extension method, the workaround of @flibustier7seas is not necessary. With that, references are wrapped in an allOf and `SupportNonNullableReferenceTypes` also applies to them.

Properties "id", "firstName", "lastName", "email" are not required with this configuration.

```csharp
services.AddSwaggerGen(builder =>
{
    builder.UseAllOfToExtendReferenceSchemas();
    ...
});
```

```json
{
  "User": {
    "type": "object",
    "properties": {
      "id": {
        "type": "string",
        "format": "uuid"
      },
      "isDeleted": {
        "type": "boolean"
      },
      "organizationId": {
        "type": "string",
        "format": "uuid",
        "nullable": true
      },
      "firstName": {
        "type": "string"
      },
      "middleName": {
        "type": "string",
        "nullable": true
      },
      "lastName": {
        "type": "string"
      },
      "email": {
        "allOf": [
          {
            "$ref": "#/components/schemas/UserEmail"
          }
        ]
      },
      "additionalEmail": {
        "allOf": [
          {
            "$ref": "#/components/schemas/UserEmail"
          }
        ],
        "nullable": true
      },
      "phone": {
        "allOf": [
          {
            "$ref": "#/components/schemas/UserPhone"
          }
        ]
      },
      "additionalPhone": {
        "allOf": [
          {
            "$ref": "#/components/schemas/UserPhone"
          }
        ],
        "nullable": true
      }
    },
    "additionalProperties": false
  }
}
```

**Comment by niemyjski** [comment]
I would love to see this work out of the box

**Comment by sebasptsch** [comment]
+1 For this feature

**Comment by Rookian** [comment]
> If you use the `SwaggerGenOptions.UseAllOfToExtendReferenceSchemas()` extension method, the `FixNullableProperties` workaround of @flibustier7seas is not necessary. With that, references are wrapped in an allOf and `SupportNonNullableReferenceTypes` also applies to them.

This won't work for `string` type or e.g. `DateTimeOffset`.

**Comment by josundt** [comment]
With C# 11.0/.NET 7.0, the `required` keyword is being introduced for properties.
Properties with this keyword needs to be initialized through the object initializer when the class/struct/record is initialized from code.

_But required properties also affects the behavior of System.Text.Json 7.x deserialization and thereby aspnetcore model validation._

The default behavior for System.Text.Json deserialization / aspnetcore model validation in .NET 7.0 is as follows.

> Nullable/not required properties:
>   - Property CAN be null and property CAN be omitted from JSON payload.

> Not nullable/not required properties:
>    - Property CAN NOT be null, but property CAN be omitted from JSON payload.

>Not nullable/required properties:
>    - Property CAN NOT be null, and property CAN NOT be omitted from JSON payload. 

Since nullablility and required/not are different things where one does not necessarily imply the other, I think SwasBuckle should use the C# property nullability for swagger property `nullable`, and C# property required/not for the swagger property `required`.

If the C# property is annotated with `RequiredAttribute`, I think this should override the above, and set both swagger nullable/required to true.

References:
* [What’s new in System.Text.Json in .NET 7 - Required Members](https://devblogs.microsoft.com/dotnet/system-text-json-in-dotnet-7/#required-members)
* [.NET Required Properties](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/required-properties)

**Comment by JohnGalt1717** [comment]
> With C# 11.0/.NET 7.0, the `required` keyword is being introduced for properties. Properties with this keyword needs to be initialized through the object initializer when the class/struct/record is initialized from code.
> 
> _But required properties also affects the behavior of System.Text.Json 7.x deserialization and thereby aspnetcore model validation._
> 
> The default behavior for System.Text.Json deserialization / aspnetcore model validation in .NET 7.0 is as follows.
> 
> > Nullable/not required properties:
> > 
> > * Property CAN be null, and property CAN be omitted from JSON payload.
> 
> > Nullable/required properties:
> > 
> > * Property CAN be null, but property CAN NOT be omitted from JSON payload.
> 
> > Not nullable/not required properties:
> > 
> > * Property CAN NOT be null, but property CAN NOT be omitted from JSON payload (since it implies null).
> 
> > Not nullable/required properties:
> > 
> > * Property CAN NOT be null, and property CAN NOT be omitted from JSON payload.
> 
> Since nullablility and required/not are different things where one does not necessarily imply the other, I think SwasBuckle should use the C# property nullability for swagger property `nullable`, and C# property required/not for the swagger property `required`.
> 
> If the C# property is annotated with `RequiredAttribute`, I think this should override the above, and set both swagger nullable/required to true.
> 
> References:
> 
> * [What’s new in System.Text.Json in .NET 7 - Required Members](https://devblogs.microsoft.com/dotnet/system-text-json-in-dotnet-7/#required-members)
> * [.NET Required Properties](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/required-properties)

... And if C# 11's required modifier is specified it should also set required regardless of if it's a string or not.

**Comment by piskov** [comment]
One more case to consider: avoiding inferred required attribute when default value is defined.

```
public string Prop { get; init; } = "default value";
```

It’s non-nullable, yet it shoudn’t be required as it has a default value.

See how [ASP validation is working](https://github.com/dotnet/aspnetcore/pull/38801) as a result of this [issue](https://github.com/dotnet/aspnetcore/issues/18403)

**Comment by JohnGalt1717** [comment]
Keep in mind that the required keyword rules everything now.

**Comment by piskov** [comment]
> Keep in mind that the required keyword rules everything now.

Absolutely not. When you use `required`, caller __must__ provide the value. For example this code will not compile even with default value:

```
var p = new Person(); 
// ↑ compile error: Required member 'Person.Name' must be set in the object initializer

  public class Person 
{ 
    public required string Name { get; init; } = "Will not work"; 
}
```

There’re a lot of use-cases for non-nullable but also not required properties. For example, some options with default values. You don’t want them to be nullable in your code (as they will never be with default value), yet you don’t want to force consumers to supply them, hence the default value:

```
  public class Project
{ 
    public SomeSettings Prop { get; init; } = new();
}
```

↑ // non-nullable, yet not required. This is exactly how ASP checker [works](https://github.com/dotnet/aspnetcore/pull/38801/files)

tldr; required non-nullable means you must provide the value; non-nullable with default value means, it’s ok to ommit as some sensible value will be used.

**Comment by JohnGalt1717** [comment]
We’re in agreement.  No required means same as before. 

Required being there means always must be filled in.

**Comment by piskov** [comment]
Yet some PRs and this sentence ↓ seem to ignore the default logic.

> Not nullable/not required properties:
> Property CAN NOT be null, but property CAN NOT be omitted from JSON payload (since it implies null).

It obviosly CAN be ommited from json IF default value is set (this way it DOES not imply null).

**Comment by josundt** [comment]
> Yet some PRs and this sentence ↓ seem to ignore the default logic.
> 
> > Not nullable/not required properties:
> > Property CAN NOT be null, but property CAN NOT be omitted from JSON payload (since it implies null).
> 
> It obviosly CAN be ommited from json IF default value is set (this way it DOES not imply null).

You're right, for this bullet there was a unintentional `NOT` in my original comment.
I updated it now.

**Comment by josundt** [comment]
@cremor
I guess `not nullable` properties i most cases should also be `required`.

But I also see that it sometimes may be useful to make `not nullable` properties optional (a.k.a. `not required`), and allow them to be omitted in JSON payloads or in object initializers.

I agree that the combination `not nullable/not required` could be a bit illogical, and arguably a bad design decision in many cases. F.ex. in a WebAPI context when a JSON payload is submitted, the API consumer may not expect that properties omitted in the payload to magically have some default values assigned on the server-side.

**Comment by piskov** [comment]
@josundtit it will be “bad design decision” only if default values are surprising to user. Suppose you `ProfileDto` and `VisibilityOptionsDto Settings`(which consists of 20 bool flags) inside it.

You have 3 options:
1. Make `VisibilityOptionsDto? Settings` nullable in C# and thus be forced to always check it for null.
2. 1. Make it required non-nullable `required VisibilityOptionsDto Settings` and force both API client and your own c# code to provide default value even if it’s just an object with 20 false bools.
3. Make it non-nullable with simple default value `VisibilityOptionsDto Settings {get; set; } = new()`. This is the best of all worlds: neither you, nor client is needed to supply an object with 20 false bools, and you are free of useless nullability checks in C# code.

So as long as default value is clear, it’s completely reasonable to support this. Imagine if you were forced to explicitly supply every property of the class every time you to create a new object in C#.

**Comment by martincostello** [comment]
https://github.com/domaindrivendev/Swashbuckle.AspNetCore/pull/2710#issuecomment-2053572112

**Comment by blouflashdb** [comment]
i am no longer able to work on it.

**Comment by martincostello** [comment]
If anyone else would like to pick up the above fix and take it forward that would be welcome.

**Comment by thomasmaisel** [comment]
Here's a updated version of the filter for .NET 10, slightly rewritten:
(mind the case insensitivity)

```csharp
/// <summary>
/// Filter to set non-nullable properties as required.
/// https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/2036
/// </summary>
internal class RequiredNotNullableSchemaFilter : ISchemaFilter
{
    public void Apply(IOpenApiSchema schema, SchemaFilterContext context)
    {
        if (schema.Properties is not { Count: > 0 }
         || schema.Required is null)
            return;

        foreach (var property in schema.Properties)
        {
            var propInfo = context.Type.GetProperty(property.Key, BindingFlags.Public | BindingFlags.Instance | BindingFlags.IgnoreCase);
            if (propInfo is null)
                continue;

            if (!IsNullableType(propInfo
             && !schema.Required.Contains(property.Key))
            {
                schema.Required.Add(property.Key);
            }
        }
    }

    public static bool IsNullableType(PropertyInfo property)
    {
        var type = property.PropertyType;

        // Handle Value types: only Nullable<T> is nullable
        if (type.IsValueType)
            return Nullable.GetUnderlyingType(type) is not null;

        // Handle Reference types: read nullable metadata (T vs T?)
        var ctx = new NullabilityInfoContext();
        var info = ctx.Create(property);
        return info.ReadState == NullabilityState.Nullable;
    }
}
```
